contentprovider.go

```diff
@@ -163,82 +173,104 @@ func (p *contentProvider) fillMatches(ms []*candidateMatch, numContextLines int,
 	return result
 }
 
-func (p *contentProvider) fillContentMatches(ms []*candidateMatch, numContextLines int) []LineMatch {
-	var result []LineMatch
-	for len(ms) > 0 {
-		m := ms[0]
-		num, lineStart, lineEnd := p.newlines().atOffset(m.byteOffset)
-
-		var lineCands []*candidateMatch
-
-		endMatch := m.byteOffset + m.byteMatchSz
-
-		for len(ms) > 0 {
-			m := ms[0]
-			if int(m.byteOffset) <= lineEnd {
-				endMatch = m.byteOffset + m.byteMatchSz
-				lineCands = append(lineCands, m)
-				ms = ms[1:]
-			} else {
-				break
+func (p *contentProvider) fillContentMatches(ms []*candidateMatch, numContextLines int) []ChunkMatch {
+	newlines := p.newlines()
+	chunks := chunkCandidates(ms, newlines, numContextLines)
+	data := p.data(false)
+	chunkMatches := make([]ChunkMatch, 0, len(chunks))
+	for _, chunk := range chunks {
+		ranges := make([]Range, 0, len(chunk.candidates))
+		for _, cm := range chunk.candidates {
+			startOffset := cm.byteOffset
+			endOffset := cm.byteOffset + cm.byteMatchSz
+			startLine, startLineOffset, _ := newlines.atOffset(startOffset)
+			endLine, endLineOffset, _ := newlines.atOffset(endOffset)
+
+			var si *Symbol
+			if cm.symbol {
+				start := p.id.fileEndSymbol[p.idx]
+				si = p.id.symbols.data(start + cm.symbolIdx)
+				if si != nil {
+					sec := p.docSections()[cm.symbolIdx]
+					si.Sym = string(data[sec.Start:sec.End])
+				}
 			}
-		}
 
-		if len(lineCands) == 0 {
-			log.Panicf(
-				"%s %v infinite loop: num %d start,end %d,%d, offset %d",
-				p.id.fileName(p.idx), p.id.metaData,
-				num, lineStart, lineEnd,
-				m.byteOffset)
-		}
-
-		data := p.data(false)
-
-		// Due to merging matches, we may have a match that
-		// crosses a line boundary. Prevent confusion by
-		// taking lines until we pass the last match
-		for lineEnd < len(data) && endMatch > uint32(lineEnd) {
-			next := bytes.IndexByte(data[lineEnd+1:], '\n')
-			if next == -1 {
-				lineEnd = len(data)
-			} else {
-				// TODO(hanwen): test that checks "+1" part here.
-				lineEnd += next + 1
-			}
+			ranges = append(ranges, Range{
+				Start: Location{
+					ByteOffset: startOffset,
+					LineNumber: uint32(startLine),
+					Column:     uint32(utf8.RuneCount(data[startLineOffset:startOffset]) + 1),
+				},
+				End: Location{
+					ByteOffset: endOffset,
+					LineNumber: uint32(endLine),
+					Column:     uint32(utf8.RuneCount(data[endLineOffset:endOffset]) + 1),
+				},
+				SymbolInfo: si,
+			})
 		}
 
-		finalMatch := LineMatch{
-			LineStart:  lineStart,
-			LineEnd:    lineEnd,
-			LineNumber: num,
+		firstLineNumber := chunk.firstLine - numContextLines
+		if firstLineNumber < 1 {
+			firstLineNumber = 1
 		}
-		finalMatch.Line = data[lineStart:lineEnd]
+		firstLineStart, _ := newlines.lineBounds(firstLineNumber)
```

Comment:
probably premature optimization, but can we avoid this lineBounds call given normally numContextLines is 0?

Reply:
We will always need to do this lookup at some point since we don't have the offset of the beginning of the line, we only have the offset of the beginning of the first range on this line.